### PR TITLE
feat: role based auth using graphback/keycloak-authz

### PR DESCRIPTION
**Pull request by darahayes**
##### Checklist

- [ ] `npm test` passes
- [ ] `npm run build` works
- [ ] tests are included
- [ ] documentation is changed or added

### CODE REVIEW

File: server/src/KeycloakCrudService.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+ */
+export class KeycloakCrudService<T = any> extends CRUDService {
+
+  private authRules: CrudServiceAuthRules;
+
+  constructor({ modelType, db, subscriptionConfig, authRules }: KeycloakCrudServiceOptions) {
+    super(modelType, db, subscriptionConfig);
+    this.authRules = authRules
+  }
+
```
Comment (darahayes) [inline server/src/KeycloakCrudService.ts:38]: This file shows how we could potentially use a CRUD service to implement auth. I was investigating this before I decided to go with the resolver wrapping approach. It's still a possibility.

File: server/src/resolvers/models.ts [hunk 2]
```diff
@@ -5,6 +5,22 @@ export const models = [
...
+          allow: 'owner',
+          operations: ['update', 'create', 'read', 'delete'],
+          ownerField: 'owner',
+        },
+        {
+          allow: 'role',
+          operations: ['update', 'delete'],
+          roles: ["admin"]
+        }
+      ]
```
Comment (darahayes) [inline server/src/resolvers/models.ts:23]: IMO this should be the output of the auth plugin. The plugin would take either directives or annotations on the model and convert it into this config here.

File: server/src/resolvers/resolvers.ts [hunk 3]
```diff
@@ -46,3 +49,9 @@ export default {
     }
   }
 }
+
+// wrap the resolvers in a seamless way such that
+// no extra code needs to be written by the user
+// this would require a very small abount of codegen but I think 
+// it's worth it.
+export default wrapResolversWithAuth(resolvers, models) as any
```
Comment (darahayes) [inline server/src/resolvers/resolvers.ts:57]: Looking for feedback on this approach

File: server/src/resolvers/wrapResolversWithAuth.ts [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+ * This has the actual auth prototype implementation
+ * 
+ * The wrapResolversWithAuth function below calls this function
+ * for each resolver
+ * 
+ * @param resolverAuthInfo the config for a specific resolver
+ * @param resolveFn the original resolver function
+ * 
+ */
+function wrapResolverWithAuth(resolverAuthInfo: any, resolveFn: any): any {
```
Comment (darahayes) [inline server/src/resolvers/wrapResolversWithAuth.ts:14]: This is the wrapping resolvers approach. Going with the wrapping approach means we should be able to follow one single approach that covers most (if not all of the auth cases), whereas going with the CrudService approach will probably require some level of wrapping in addition.

There is a benefit of this approach - it's fairly easy to understand and follow.

The big drawback I see is that wrapping resolvers is a little weird and might not play nicely if the user wants to do some of their own wrapping. Looking for feedback on this.

File: server/src/resolvers/wrapResolversWithAuth.ts [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+/**
+ * 
+ * @param resolversMap Resolver functions for your application
+ * @param authInfo the auth config for the resolvers
+ * 
+ * Wraps the resolvers with a wrapper function that can perform auth checks
+ * based on some config passed in
+ */
+export function wrapResolversWithAuth(resolversMap: any, models: any) {
```
Comment (darahayes) [inline server/src/resolvers/wrapResolversWithAuth.ts:65]: This does the actual wrapping for all resolvers. The rest of the code in this file is some low level messy object manipulation stuff

File: server/src/KeycloakCrudService.ts [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+  read: CrudOperationAuthRules,
+  update: CrudOperationAuthRules,
+  delete: CrudOperationAuthRules,
+}
+
+export type KeycloakCrudServiceOptions = {
+  modelType: GraphQLObjectType,
+  db: GraphbackDataProvider,
+  subscriptionConfig: GraphbackPubSub,
+  authRules: CrudServiceAuthRules,
```
Comment (wtrocki) [inline server/src/KeycloakCrudService.ts:21]: I love how nice this fits to graphback - models.ts can contain representation of the auth rules

File: server/src/resolvers/models.ts [hunk 7]
```diff
@@ -5,6 +5,22 @@ export const models = [
       publishCreate: true,
       publishUpdate: true,
       publishDelete: true
+    },
+    // this auth config piece should be generated by the graphback plugin
+    // we can discuss whether that's directives or annotations
+    auth: {
```
Comment (wtrocki) [inline server/src/resolvers/models.ts:11]: It will be annotations. The only deal here is that his will be convoluted in terms of packages:

- Generator package will be required in devDependencies
- Service package will be required in dependencies to implement KeycloakService

Comment (wtrocki) [reply server/src/resolvers/models.ts:11]: Challenge to compose this: Resolves plugin will generate models but then separate plugin will need to overwrite it in some specific way.

Comment (darahayes) [reply server/src/resolvers/models.ts:11]: Yeah. I think that something similar should be done with how the schema is handled, it would be cool if you could pass the models through the plugins before they get written

**Comment by wtrocki** [comment]
Main question: if we imagine that resolves will not be generated but always available as code - would this affect in any way this investigation/auth implementations? How many issues we will resolve by removing generation towards only runtime. 

@darahayes, In this case, our composition could look like pre/post wrappers around graphback DataProviders. This will invalidate need for service abstraction (which is not really flexible as requires the creation of the class instead of single-function override.

The runtime will mean that - people get resolvers as a variable (in memory implementation) - the same way as they do in generated code but they do not need to import resolvers or have physical file doing some boilerplate work around queries. 

@darahayes @craicoverflow What do you think about this?

I do feel that there are no positives of generating code apart from possible slow startup times required to build runtime resolves.
There is still value in generating schema and client-side queries, but resolvers seem to be limiting us from having better getting started experience and 
This will affect both CRUD spec work (minimize the scope of both implementations), minimize the complexity of the code. Simplify ways to test solutions, with little to no drawbacks overall. 

What will happen with the resolver generator plugin? 
We will deprecate it and use the already implemented resolver layer from runtime?

**Comment by wtrocki** [comment]
The way I see us driving this is to interpret schema at runtime with provided annotations (or codegen if needed) and have pre post predicates (separate) for specific use cases (role, owner etc.)

**Comment by wtrocki** [comment]
Entire problem with current approach is that we provided this service abstraction that is somehow recommended as way to extend resolvers where there is already much simpler and easier way to do it

**Comment by darahayes** [comment]
> Main question: if we imagine that resolves will not be generated but always available as code - would this affect in any way this investigation/auth implementations? How many issues we will resolve by removing generation towards only runtime.

While they are somewhat related, I think the auth work can proceed regardless of this. If we go with some version of the resolver wrapping approach, (which makes the most sense to me). Then it makes no difference where the original resolvers come from because the wrapper bit and pieces do not care.

> In this case, our composition could look like pre/post wrappers around graphback DataProviders.

This is just another way of describing resolver wrapping or 'middlewares' so I think we're on the same page.

> Entire problem with current approach is that we provided this service abstraction that is somehow recommended as way to extend resolvers where there is already much simpler and easier way to do it

It depends on how you look at it. The service abstraction is a fantastic abstraction for a CRUD model that is actually not aware of GraphQL at all. When you extend the CRUD service you'e not really modifying how resolvers behave. You're modifying how the CRUD operations behave. However, when you want to do things that are actually GraphQL aware, then you have to go outside of that abstraction. An example would be the auth plugin wrapping default field resolvers on scalars (such as a protected `User.ssn` resolver)

> The way I see us driving security is to interpret schema at runtime with provided annotations (or codegen if needed) and have pre-post predicates (separate) for specific use cases (role, owner etc.)

Again I think this is just another way of describing resolver wrappers/middlewares. When you talk about interpreting the schema at runtime with provided annotations, do you mean having annotations or directives on the runtime schema? Because I don't think it's necessary. Implementing directives is actually pretty tricky, implementing directives in a way that it works both on Apollo Server/`graphql-tools` and graphql-js sounds like way more trouble than it's worth. All we need is to output some config as shown in the PR and then pick those up at runtime in your pre/post resolver predicates. This approach is much simpler and will allow for future extensions in a reliable way.

For example, today we might be only generating some auth rules and outputting them in a JSON file, but in the future we could be inspecting those rules and ensuring that the roles and other resources are actually created in a live keycloak instance. We could be generating full infrastructure as code templates to actually provision the application and configure external services that the application depends on. All of it would be checked into your git repo and live alongside the application. I see other existing codegen tools doing similar things.

So TL;DR

* If you feel strongly that the service abstraction is not needed, I can get behind that idea and it will to affect this investigation too much.
* We can go with resolver wrappers or middlewares or pre/post predicates or whatever you want to call them 🙂
* I think Graphback plugins should be able to generate and extend config files as shown in the `models.ts` file in this PR. And I think in future that same approach will lead to things like external service configurations or auto generated infrastructure as code.

**Comment by darahayes** [comment]
Ping @wtrocki this PR has been updated to use the newly published library.

**Comment by darahayes** [comment]
Auth package PR https://github.com/aerogear/graphback/pull/1255

**Comment by wtrocki** [comment]
Yep

File: server/src/createContext.ts [hunk 8]
```diff
@@ -36,9 +40,15 @@ export const createOffixMongoCRUDRuntimeContext = (
...
-            ...model.pubSub
+
+        services[model.name] = new AMQCRUDService({
+            modelType,
+            db: objectDB,
+            subscriptionConfig: {
+                pubSub,
+                ...model.pubSub
+            },
+            authConfig: serviceAuthConfigs[model.name]
```
Comment (wtrocki) [inline server/src/createContext.ts:51]: As in other PR - what will be the reasoning for this approach

File: server/src/createContext.ts [hunk 9]
```diff
@@ -36,9 +38,15 @@ export const createOffixMongoCRUDRuntimeContext = (
...
-            ...model.pubSub
+
+        services[model.name] = new AMQCRUDService({
+            modelType,
+            db: objectDB,
+            subscriptionConfig: {
+                pubSub,
+                ...model.pubSub
+            },
+            authConfig: authConfig[model.name]
```
Comment (darahayes) [inline server/src/createContext.ts:49]: @wtrocki trying to figure out how to document the usage of the library.

This stuff is happening inside a helper function that we copy pasted into the application. Normally the user doesn't really write the code to create all those services.

I can write a helper function in the original library that looks like this one but accepts a DataProvider constructor. That way the helper can create everything fairly easily. All of the existing DataProvider classes in the Graphback Project currently take the same args but it's not actually defined which means it's not really guaranteed. Any ideas?

The API here for setting up the runtime is getting quite convoluted. I had a PR to try address it but it was closed.

Comment (wtrocki) [reply server/src/createContext.ts:49]: Please connect this any way without spending much time. 
Helper works great for me at the moment.
We will refine it once this lands.

File: server/src/graphql.ts [hunk 10]
```diff
@@ -20,7 +23,16 @@ export const createApolloServer = async function (app: Express, config: Config)
 
     const typeDefs = loadSchemaFiles(join(__dirname, '/schema/')).join('\n');
     const schema = buildSchema(typeDefs, { assumeValid: true });
-    const context = createOffixMongoCRUDRuntimeContext(models, schema, db, pubSub);
+
+    const context = createKeycloakRuntimeContext({
```
Comment (darahayes) [inline server/src/graphql.ts:27]: cc @wtrocki I added this helper method to the auth package. It's kinda based on the idea from a previous PR that was doing some refactors to the Graphback runtime. It allows you to pass in the uninitialised DataProvider and CrudService class and the helper will create them for you. You can see the auth package PR for how it works. What do you think of this approach?


Going with an approach like this potentially removes the need for different helper functions for all the different DataProvider classes given they all do the exact same thing. 

To make it really solid I think we'd have to refactor the constructors a little bit in the existing DataProvider and CrudService classes to accept a single options object (that can be extended) instead of multiple arguments.

Comment (wtrocki) [reply server/src/graphql.ts:27]: > To make it really solid I think we'd have to refactor the constructors a little bit in the existing DataProvider and CrudService classes to accept a single options object (that can be extended) instead of multiple arguments.

I would love to do it with you but we want to deliver this feature first and I do not see this as blocker. I have a couple issues to resolve before we need to go with. Since we are working on CRUDspec now on feature branch it will be unwise to start refactoring this. 

PR's that are blocking:
CRUD feature branch
Issues that are blocking this
https://github.com/aerogear/graphback/issues/1159
https://github.com/aerogear/graphback/issues/1157
https://github.com/aerogear/graphback/issues/1158

Comment (wtrocki) [reply server/src/graphql.ts:27]: We not going to get this solid until we try this out in the regular use cases so we are happy to get something out there - this will be alpha anyway.

File: client/src/graphql/fragments/Task.ts [hunk 11]
```diff
@@ -6,6 +6,7 @@ export const TaskFragment = gql`
    title
    description
    status
+   owner
```
Comment (wtrocki) [inline client/src/graphql/fragments/Task.ts:9]: To be removed

File: server/src/config/_keycloak.json [hunk 12]
```diff
@@ -1,9 +1,9 @@
 {
   "realm": "voyager-testing",
+  "bearer-only": true,
```
Comment (wtrocki) [inline server/src/config/_keycloak.json:3]: We need to change scripts then as we print both public ones (same in WT)

Comment (wtrocki) [reply server/src/config/_keycloak.json:3]: Or revert it

File: server/integrations/keycloak/initKeycloak.js [hunk 13]
```diff
@@ -116,7 +118,8 @@ async function prepareKeycloak() {
 
       // Assign roles to the user
       await assignRealmRolesToUser(user, userIdUrl)
-      await assignClientRolesToUser(user, BEARER_CLIENT, userIdUrl)
+      await assignClientRolesToUser(user, BEARER_CLIENT, bearerClientRoles, userIdUrl)
+      await assignClientRolesToUser(user, PUBLIC_CLIENT, publicClientRoles, userIdUrl)
```
Comment (darahayes) [inline server/integrations/keycloak/initKeycloak.js:122]: basically the helper script used to assign some roles to the user only on the bearer client but because we use the public client in the backend also in the walkthrough etc, then the role checks are against the public client. This change makes sure the client is assigned the roles on both the public and bearer client so the role checks will work regardless of which client is used.

**Comment by darahayes** [comment]
@wtrocki ready for final review

File: server/integrations/keycloak/initKeycloak.js (hunk 13 above)
Comment (wtrocki) [reply server/integrations/keycloak/initKeycloak.js:122]: [Question] Could we just export the realm json with this roles?

**Comment by wtrocki** [comment]
What kind of user I should be using here? I do not see actual effects of this change

**Comment by wtrocki** [comment]
It seems that UI has no indication of the lack of permissions which can be confusing and might bring some bug reports

**Comment by wtrocki** [comment]
I'm happy to get that into starter but there are some areas to improve

File: client/src/components/TaskList.tsx [hunk 14]
```diff
@@ -54,6 +59,14 @@ export const TaskList: React.FC<any> = ({ tasks }) => {
           })
         }
       </IonList>
+      <IonToast
```
Comment (darahayes) [inline client/src/components/TaskList.tsx:62]: @wtrocki these changes now show an error for deleting tasks and an error also gets shown for updating tasks when the user is not authorized.

**Comment by darahayes** [comment]
ready for final review

File: server/integrations/keycloak/initKeycloak.js (hunk 13 above)
Comment (wtrocki) [reply server/integrations/keycloak/initKeycloak.js:122]: Repeating question

File: client/src/components/TaskList.tsx (hunk 14 above)
Comment (wtrocki) [reply client/src/components/TaskList.tsx:62]: Yes. Sorry for ignorance - I forgot that we have errors sent already sent from server
Dialog is an actual bonus. Good work

File: server/integrations/keycloak/initKeycloak.js (hunk 13 above)
Comment (darahayes) [reply server/integrations/keycloak/initKeycloak.js:122]: Yeah I'm pretty sure it's not possible. Any time I've tried before it didn't really work for me. This script is a little tricky to maintain because of all the handcrafted axios requests going to keycloak, but it could be replaced with a much simpler version of the same script we have in `keycloak-connect-graphql` https://github.com/aerogear/keycloak-connect-graphql/blob/master/scripts/initKeycloak.js
